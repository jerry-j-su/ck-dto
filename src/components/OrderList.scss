@use 'sass:color';
@use '../theme.scss';

.order-list-container {
    &.hidden {
        // visibility: hidden!important;
        // height: 0!important;
        display:none;
    }

    height: 100%;
    overflow: hidden auto;
    scrollbar-gutter: stable both-edges; // very important, prevent visible scroll bar causing inner content to reflow/repaint
    &::-webkit-scrollbar {
        // display: none;
        width: 5px;
        height: 5px;
    }
    &::-webkit-scrollbar-track {
        border-radius: 2px;
        background-color: color.scale(theme.$bg-front, $whiteness: 7%);
    }
    &::-webkit-scrollbar-thumb {
        border-radius: 2px;
        background-color: theme.$font-primary-strong;
    }

    .order-count {
        font-size: .8rem;
        color: theme.$font-fade;
    }

    .order-list-body {
        padding: 0;
        margin-block: 0;
        list-style-type: none;
        display: grid;
        flex-direction: column;
    }

    .order-wrapper {
        border: 1px solid color.scale(theme.$border-fade, $blackness: 30%);
        border-radius: .2px;
        margin: .2rem;

        dl {
            display: flex;
            flex-wrap: wrap;
            margin: 0;
            padding: .5rem 0;

            .top-row,
            .col-1,
            .col-2 {
                padding: .2rem .5rem;
            }

            .top-row {
                flex: 1 0 100%;
            }

            .col-1 {
                flex: 0 1 15%;
                display: flex;
                justify-content: center;
                align-items: center;
            }

            .col-2 {
                flex: 1 0 auto;
            }

            .label {
                font-size: .6rem;
                color: theme.$font-fade;
                vertical-align: middle;
            }

            .value {
                &.name {
                    font-weight: 700;
                }

                &.status,
                &.items {
                    font-size: .8rem;
                }

                &.name {
                    color: theme.$font-primary-strong;
                }

                &.price {
                    color: theme.$highlight;
                }
            }

            dt {
                display: inline;
            }

            dd {
                display: inline;
                margin-inline-start: .5rem;
                margin-bottom: .2rem;
            }

            .name-spacer {
                flex: 1 0 90%;
                border-bottom: 1px dashed color.scale(theme.$border-fade, $blackness: 50%);
                margin: 0 1rem 0.2rem 1rem;
            }

            .status-spacer {
                flex: 0 0 auto;
                border-left: 1px dashed color.scale(theme.$border-fade, $blackness: 50%);
                margin: .7rem 0;
            }

            .spacer {
                margin-block: .5rem;
            }
        }
    }
}

.scrollable-wrapper {
    width: 100%;
    min-height: 0;
    position: relative;
    .scrollable-tip {
        &-top {
            position: absolute;
            top: 0;
        }
        &-bottom {
            position: absolute;
            bottom: 0;
        }
    }
}
