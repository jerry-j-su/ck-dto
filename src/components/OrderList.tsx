import React, { useState, useEffect, useMemo, forwardRef } from 'react'

import { OrderType } from '../types'
import { ScrollableContainerExtraProps } from './dynamicScrollableListHOC'
import OrderSkeleton from './OrderSkeleton'
import { mergeClassNames } from '../utils'
import './OrderList.scss'

type OrderListProps = {
    orders: OrderType[],
    containerClassName?: string,
}
const UpdateBatchSize = 10  // how many buffered records to attach to DOM tree if user scroll down to the bottom of the page

export default forwardRef(({ orders, containerClassName, holdAppendChild }: OrderListProps & ScrollableContainerExtraProps, ref: any) => {
    const [renderCount, setRenderCount] = useState<number>(0)
    useEffect(() => {
        if (holdAppendChild) return
        if (renderCount < orders.length) {
            setRenderCount(renderCount + UpdateBatchSize)
        }
    }, [holdAppendChild, orders, orders.length, renderCount, setRenderCount])

    const ordersToShow = useMemo(() => {
        return orders.slice(0, renderCount)
    }, [orders, orders.length, renderCount]) // eslint-disable-line react-hooks/exhaustive-deps

    return (
        <section className={mergeClassNames(['order-list-container', containerClassName])} ref={ref} tabIndex={0}>
            <ul className="order-list-body">
                {ordersToShow.map(({ id, customer, destination, 'event_name': status, item, price, 'sent_at_second': time }) => (
                    <li className="order-wrapper" key={`${id}-${status}`}>
                            <div className="top-row">
                                <span className="label name">name</span>
                                <span className="value name">{customer}</span>
                            </div>
                            <div className="name-spacer" />
                            <div className="col-1">
                                <span className="label status" style={{ display: 'none' }}>Status</span>
                                <span className="value status">{status}</span>
                            </div>
                            <div className="status-spacer" />
                            <div className="col-2">
                                <span className="label items">items</span>
                                <span className="value items">{item}</span>
                                <div className="spacer" />
                                <span className="label price">price</span>
                                <span className="value price">{price}</span>
                            </div>
                    </li>
                ))}
                {/* DOM append is held due, display skeleton */}
                {holdAppendChild && renderCount < orders.length
                    ? (<>
                        <li className="order-wrapper skeleton"><OrderSkeleton /></li>
                        <li className="order-wrapper skeleton"><OrderSkeleton /></li>
                      </>)
                    : <li className="order-wrapper end"><div className="loading-dot" /></li>
                }
            </ul>
        </section>
    )
})
