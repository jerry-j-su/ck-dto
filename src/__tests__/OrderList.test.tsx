import React from 'react'
import { readFileSync } from 'fs'

import OrderList from '../components/OrderList'


// module mocks
const dummyRenderCount = 1
const mockSetRenderFunc = jest.fn()
const sideEffectList: Function[] = []
jest.mock('react', () => {
    const originalModule = jest.requireActual('react')
    return {
        __esModule: true,
        ...originalModule,
        forwardRef: (cb: Function) => cb,
        useState: (initialValue: any) => [dummyRenderCount, mockSetRenderFunc],  // for the renderCount useState
        useEffect: (cb: Function) => sideEffectList.push(cb),
        useMemo: (cb: Function) => cb(),
    }
})

// dummy data
const mockOrders = JSON.parse(readFileSync(`${__dirname}/dummyOrders.json`, { encoding: 'utf-8' })).slice(1)

describe('<OrderList> Render tests', () => {
    beforeEach(() => {
        mockSetRenderFunc.mockClear()
        sideEffectList.splice(0)    // clear the side effect function list
    })

    test('Basic render', () => {
        const result = OrderList({ orders: mockOrders, containerClassName: 'dummy-container' })
        expect(result).toMatchSnapshot()
    })

    test('holdAppendChild not set, if render account is less than order count, continue adding order record DOM', () => {
        OrderList({ orders: mockOrders, containerClassName: 'dummy-container' })
        expect(mockSetRenderFunc).not.toHaveBeenCalled()
        const sideEffect = sideEffectList.pop()
        if (sideEffect) sideEffect()
        expect(mockSetRenderFunc).toHaveBeenCalled()
    })

    test('holdAppendChild is set, order record rendering is on hold, render count stay the same', () => {
        OrderList({ orders: mockOrders, containerClassName: 'dummy-container', holdAppendChild: true })
        expect(mockSetRenderFunc).not.toHaveBeenCalled()
        const sideEffect = sideEffectList.pop()
        if (sideEffect) sideEffect()
        expect(mockSetRenderFunc).not.toHaveBeenCalled()
    })
})
