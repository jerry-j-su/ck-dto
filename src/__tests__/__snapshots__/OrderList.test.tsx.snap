// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderList> Render tests Basic render 1`] = `
<section
  className="order-list-container dummy-container"
  tabIndex={0}
>
  <ul
    className="order-list-body"
  >
    <li
      className="order-wrapper"
    >
      <div
        className="top-row"
      >
        <span
          className="label name"
        >
          name
        </span>
        <span
          className="value name"
        >
          Derek Marshall
        </span>
      </div>
      <div
        className="name-spacer"
      />
      <div
        className="col-1"
      >
        <span
          className="label status"
          style={
            Object {
              "display": "none",
            }
          }
        >
          Status
        </span>
        <span
          className="value status"
        >
          DELIVERED
        </span>
      </div>
      <div
        className="status-spacer"
      />
      <div
        className="col-2"
      >
        <span
          className="label items"
        >
          items
        </span>
        <span
          className="value items"
        >
          Orange chicken
        </span>
        <div
          className="spacer"
        />
        <span
          className="label price"
        >
          price
        </span>
        <span
          className="value price"
        >
          4775
        </span>
      </div>
    </li>
    <li
      className="order-wrapper end"
    >
      <div
        className="loading-dot"
      />
    </li>
  </ul>
</section>
`;
